### PR TITLE
Optimization for getting value from map, arrayElement(2/2)

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...
This PR follows #55929, it will bring about 30% speedup.
- reduce the reserved memory
- reduce the `resize` call 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 77329846189c8c1c33a68ceec09e4f59ead11122 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55957/77329846189c8c1c33a68ceec09e4f59ead11122/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6622223231">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55957/77329846189c8c1c33a68ceec09e4f59ead11122/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by lgbo-ustc** [comment]
@antaljanosbenjamin will you mark this `can be tested` ?

**Comment by lgbo-ustc** [comment]
perf result
image

**Comment by lgbo-ustc** [comment]
@antaljanosbenjamin Do you have time to view this PR ?

**Comment by antaljanosbenjamin** [comment]
Hi! Sorry, I was/am OOO. I will check this PR on Monday.

File: src/Functions/array/arrayElement.cpp [hunk 1]
```diff
@@ -258,7 +262,7 @@ struct ArrayElementNumImpl
 
 struct ArrayElementStringImpl
 {
-    template <bool negative>
+    template <bool negative, bool used_builder>
```
Comment (antaljanosbenjamin) [inline src/Functions/array/arrayElement.cpp:265]: My only concern is this extra template parameter. Why is it better then using the passed-in `builder` object? As far as I see we can avoid some `if (builder)` checks in runtime, but I am not sure if the extra complexity worth the gains.

**Comment by antaljanosbenjamin** [comment]
Test failures shouldn't be connected to the PR:
- Upgrade check: #56202
- Stateless tests: I am investigating it

Regarding the perf results, I think the results are kind of flaky, [e.g.](https://s3.amazonaws.com/clickhouse-test-reports/55957/7f3e400e63ae43dcd90a746c8afecff313067493/performance_comparison_[2_4]/report.html):
image

So I am not convinced that the optimization actually makes ClickHouse faster. I will try to ask around how we should interpret these results properly.

**Comment by lgbo-ustc** [comment]
> Test failures shouldn't be connected to the PR:
> 
> * Upgrade check: [Logical error: Invalid number of rows in Chunk column UInt64 position 4: expected 8192, got 2 (`_block_number`) #56202](https://github.com/ClickHouse/ClickHouse/issues/56202)
> * Stateless tests: I am investigating it
> 
> Regarding the perf results, I think the results are kind of flaky, [e.g.](https://s3.amazonaws.com/clickhouse-test-reports/55957/7f3e400e63ae43dcd90a746c8afecff313067493/performance_comparison_%5B2_4%5D/report.html): image
> 
> So I am not convinced that the optimization actually makes ClickHouse faster. I will try to ask around how we should interpret these results properly.

The performance result isn't stable in CI. We  steadily improve locally。

File: src/Functions/array/arrayElement.cpp (hunk 1 above)
Comment (antaljanosbenjamin) [reply src/Functions/array/arrayElement.cpp:265]: @lgbo-ustc what do you think about this?

Comment (lgbo-ustc) [reply src/Functions/array/arrayElement.cpp:265]: Just want to reduce the branch jumps in a loop.

**Review by antaljanosbenjamin — APPROVED** [review]

**Pull request merged**